File: lib/src/application/flow/overview.dart
import 'package:functionally/common.dart';
import 'package:functionally/reader.dart' as R;
import 'package:functionally/reader_stream.dart' as RS;
import 'package:functionally/stream.dart';
import 'package:rxdart/rxdart.dart';

import '../../core/controller.dart';
import '../../domain/food.dart';
import '../commands/delete_foods_by_ids.dart';
import '../commands/foods.dart';
import '../commands/log.dart';

sealed class OverviewModel {}

final class Loading implements OverviewModel {
  const Loading();
}

final class Error implements OverviewModel {
  const Error(this.message);
  final String message;
}

final class FoodModel {
  const FoodModel({required this.name});
  final String name;
}

final class Ready implements OverviewModel {
  const Ready({
    required this.foods,
    required this.pending,
  });

  Ready.fromData({
    required Iterable<Food> foods,
    required this.pending,
  }) : foods = _listToView(foods);

  final int pending;
  final Iterable<FoodModel> foods;
}

sealed class Command {
  const Command();
}

final class Refresh implements Command {
  const Refresh(this.page);
  final int page;
}

final class Delete implements Command {
  const Delete(this.ids);

  final Set<String> ids;
}

typedef OverviewController = Controller<Command, OverviewModel>;

typedef OverviewControllerDeps = ({
  DeleteFoodsByIds deleteByIds,
  Log log,
  Foods foods,
});

final R.Reader<OverviewControllerDeps, OverviewController> overviewController =
    Builder(R.ask<OverviewControllerDeps>())
        .transform(
          R.map_(
            (_) => (
              Builder(
                R.asks(
                  (OverviewControllerDeps deps) => deps.foods,
                ),
              )
                  .transform(
                    R.flatMap(
                      (stream) => (deps) => stream.doOnListen(
                            deps.log(
                              LogType.info,
                              'Started listening to foods',
                            ),
                          ),
                    ),
                  )
                  .transform(
                    R.flatMap(
                      (stream) => (deps) => stream.doOnData(
                            (foods) => deps.log(
                              LogType.info,
                              'Received ${foods.length} food entries',
                            ),
                          ),
                    ),
                  )
                  .transform(
                    RS.map_(
                      (foods) => foods.map(toFoodEntity),
                    ),
                  )
                  .transform(
                    RS.map_(
                      // ignore: unnecessary_cast
                      (foods) => Ready.fromData(
                        foods: foods,
                        pending: 0,
                      ) as OverviewModel,
                    ),
                  )
                  .transform(
                    R.map_(
                      (stream) => stream.startWith(const Loading()),
                    ),
                  )
                  .build(),
              (Stream<Command> command$) =>
                  Builder(RS.ask<OverviewControllerDeps>())
                      .transform(
                        R.map_(
                          (_) =>
                              command$.asBroadcastStream().whereType<Delete>(),
                        ),
                      )
                      .transform(
                        R.flatMap(
                          (stream) => (deps) => stream.doOnData(
                                (_) => deps.log(
                                  LogType.info,
                                  'Received delete command',
                                ),
                              ),
                        ),
                      )
                      .transform(
                        R.flatMap(
                          (stream) => (deps) => stream.flatMap(
                                (delete) => fromTask(
                                  deps.deleteByIds(delete.ids),
                                ),
                              ),
                        ),
                      )
                      .transform(
                        R.flatMap(
                          (stream) => (deps) => stream.doOnData(
                                (_) => deps.log(
                                  LogType.info,
                                  'Delete command executed',
                                ),
                              ),
                        ),
                      )
                      .transform(
                        R.map_((stream) => stream.ignoreElements()),
                      )
                      .build(),
            ),
          ),
        )
        .transform(
          R.flatMap(
            (streams) => (env) => Controller.withPublishSubject(
                  (Stream<Command> command$) => MergeStream([
                    streams.$1(env),
                    streams.$2(command$)(env),
                  ]),
                ),
          ),
        )
        .build();

Iterable<FoodModel> _listToView(Iterable<Food> foods) =>
    foods.map((food) => FoodModel(name: food.name));

File: src/app/controllers/overview.ts
import {
	eq as Eq,
	option as OPT,
	ord as Ord,
	reader as R,
	readonlySet as RoS,
} from 'fp-ts'
import { Reader } from 'fp-ts/Reader'
import { flip, flow, pipe } from 'fp-ts/function'
import * as Rx from 'rxjs'

import { Controller } from '@/core/controller'
import * as RO from '@/core/reader-observable'
import { filterMap } from '@/core/rx'

import { Food, areEqual } from '@/domain/food'

import {
	AddFailure,
	EnqueueProcess,
	Log,
} from '@/app/actions/commands'
import {
	OnChangeProcesses,
	OnFoods,
} from '@/app/actions/streams'
import { toFoodEntity } from '@/app/types/food'
import { info } from '@/app/types/log'
import {
	Process,
	Processes,
	processesEq,
} from '@/app/types/process'

interface UseCases {
	readonly processes$: OnChangeProcesses
	readonly enqueueProcess: EnqueueProcess
	readonly foods$: OnFoods
	readonly log: Log
	readonly addFailure: AddFailure
}

interface FoodModel {
	readonly id: string
	readonly name: string
	readonly deleting: boolean
}

const foodModelEq: Eq.Eq<FoodModel> =
	Eq.fromEquals((a, b) => a.id === b.id)

export type OverviewModel = Readonly<
	| {
			type: 'ready'
			foods: readonly FoodModel[]
	  }
	| {
			type: 'loading'
	  }
>

export type Command = Readonly<{
	type: 'delete'
	ids: ReadonlySet<string>
}>

export type OverviewController = Controller<
	Command,
	OverviewModel
>

type Overview = Reader<
	UseCases,
	Controller<Command, OverviewModel>
>
export const overview: Overview = pipe(
	[
		pipe(
			R.asks(({ foods$ }: UseCases) => foods$),
			RO.tap(
				foods =>
					({ log }) =>
						pipe(
							info(
								`Received ${foods.size} food entries`,
							),
							log,
						),
			),
			RO.map(
				// We convert all food data into food entities in order to enforce business constraints.
				RoS.map(Eq.fromEquals(areEqual))(
					toFoodEntity,
				),
			),
			R.chain(
				flip(({ processes$ }) =>
					Rx.combineLatestWith(processes$),
				),
			),
			RO.map(([foods, processes]) =>
				// We create the food model by merging the entity with the queued processes
				RoS.map(foodModelEq)((food: Food) =>
					toFoodModel(food, processes),
				)(foods),
			),
			RO.map(
				RoS.toReadonlyArray(
					Ord.fromCompare(() => 0),
				),
			),
			RO.map(
				foods =>
					({
						foods,
						type: 'ready',
					}) satisfies OverviewModel,
			),
			R.map(
				Rx.startWith({
					type: 'loading',
				} satisfies OverviewModel),
			),
		),
		flow(
			R.of<UseCases, Rx.Observable<Command>>,
			RO.tap(
				() =>
					({ log }) =>
						pipe(
							info(`Received delete command`),
							log,
						),
			),
			RO.map(
				del =>
					({
						type: 'delete',
						ids: del.ids,
					}) as const,
			),
			RO.mergeMap(
				flip(({ enqueueProcess }) =>
					flow(enqueueProcess, Rx.defer),
				),
			),
			RO.tap(
				() =>
					({ log }) =>
						pipe(
							info(`Delete command executed`),
							log,
						),
			),
			R.map(filterMap(OPT.getLeft)),
			RO.tap(
				error =>
					({ log }) =>
						pipe(
							info(
								`Delete command failed: ${error.message}`,
							),
							log,
						),
			),
			RO.mergeMap(
				error =>
					({ addFailure }) =>
						pipe(addFailure(error), Rx.defer),
			),
			R.map(Rx.ignoreElements()),
		),
	] as const,

	([foods, deleteByIds]) =>
		useCases =>
			new Controller(
				cmd$ =>
					Rx.merge(
						foods(useCases),
						deleteByIds(cmd$)(useCases),
					),
				{
					type: 'loading',
				} satisfies OverviewModel,
			),
)

function toFoodModel(
	food: Food,
	processes: ReadonlySet<Process>,
): FoodModel {
	return {
		...food,
		deleting: pipe(
			processes,
			RoS.filter(
				process => process.type === 'delete',
			),
			RoS.some(
				process =>
					process.ids === food.id ||
					(!(typeof process.ids === 'string') &&
						process.ids.has(food.id)),
			),
		),
	}
}
